File: java/MyHashMapApp.java
import java.util.LinkedList;
import java.util.List;
import java.util.ArrayList;

class Pair<K, V> {
    public K first;
    public V second;

    public Pair(K first, V second) {
        this.first = first;
        this.second = second;
    }
}

class Bucket {
    private List<Pair<Integer, Integer>> bucket;

    public Bucket() {
        this.bucket = new LinkedList<Pair<Integer, Integer>>();
    }

    public Integer get(Integer key) {
        for(Pair<Integer, Integer> pair: this.bucket) {
            if(pair.first.equals(key))
                return pair.second;
        }
        return -1;
    }

    public void update(Integer key, Integer value) {
        boolean found = false;
        for(Pair<Integer, Integer> pair: this.bucket) {
            if(pair.first.equals(key)) {
                pair.second = value;
                found = true;
            }
        }
        if (!found)
            this.bucket.add(new Pair<Integer, Integer>(key, value));
    }

    public void remove(Integer key) {
        for(Pair<Integer, Integer> pair: this.bucket) {
            if(pair.first.equals(key)) {
                this.bucket.remove(pair);
                break;
            }
        }
    }
}

class MyHashMap {
    private int keySpace;
    private List<Bucket> hashTable;

    public MyHashMap() {
        this.keySpace = 2069;
        this.hashTable = new ArrayList<Bucket>();
        for(int i = 0; i < this.keySpace; i++) {
            this.hashTable.add(new Bucket());
        }
    }

    public void put(int key, int value) {
        int hashKey = key % this.keySpace;
        this.hashTable.get(hashKey).update(key, value);
    }

    public int get(int key) {
        int hashKey = key % this.keySpace;
        return this.hashTable.get(hashKey).get(key);
    }

    public void remove(int key) {
        int hashKey = key % this.keySpace;
        this.hashTable.get(hashKey).remove(key);
    }
}

class MyHashMapApp {
    public static void main(String...args) {
        MyHashMap hashMap = new MyHashMap();
        hashMap.put(5,5);
        hashMap.put(10,10);
        System.out.println(hashMap.get(5));
        System.out.println(hashMap.get(15));
        hashMap.put(10,5);
        System.out.println(hashMap.get(10));
        hashMap.remove(10);
        System.out.println(hashMap.get(10));
    }
}

File: python/myhashmapapp.py
class Bucket:
    def __init__(self):
        self.bucket = []

    def get(self, key):
        for (k, v) in self.bucket:
            if k == key:
                return v
        return -1

    def update(self, key, value):
        found = False
        for i, kv in enumerate(self.bucket):
            if key == kv[0]:
                self.bucket[i] = (key, value)
                found = True
                break
        if not found:
            self.bucket.append((key, value))

    def remove(self, key):
        for i, kv in enumerate(self.bucket):
            if key == kv[0]:
                del self.bucket[i]


class MyHashMap(object):
    def __init__(self):
        self.keySpace = 2069
        self.hashTable = [Bucket() for i in range(self.keySpace)]

    def put(self, key, value):
        hash_key = key % self.keySpace
        self.hashTable[hash_key].update(key, value)

    def get(self, key):
        hash_key = key % self.keySpace
        return self.hashTable[hash_key].get(key)

    def remove(self, key):
        hash_key = key % self.keySpace
        self.hashTable[hash_key].remove(key)


def main():
    hash_map = MyHashMap()
    hash_map.put(5, 5)
    hash_map.put(10, 10)
    print(hash_map.get(5))
    print(hash_map.get(15))
    hash_map.put(10, 5)
    print(hash_map.get(10))
    hash_map.remove(10)
    print(hash_map.get(10))


if __name__ == "__main__":
    main()
